File: ternary-search/codeforces-578-c/solution.py
from sys import stdin
from typing import Iterable

EPS = 1e-6


def poorness(seq: Iterable[float]) -> float:
    first = next(seq)  # type: ignore
    running_sum_pos = first
    running_sum_neg = first
    global_max = abs(first)

    for item in seq:
        running_sum_pos = max(item, item + running_sum_pos)
        running_sum_neg = min(item, item + running_sum_neg)
        global_max = max(global_max, running_sum_pos, abs(running_sum_neg))

    return global_max


def weakness(x: float, seq: list[int]) -> float:
    return poorness(map(lambda v: v - x, seq))


def values_and_error(
    m1: float, m2: float, values: list[int]
) -> tuple[float, float, float]:

    w1 = weakness(m1, values)
    w2 = weakness(m2, values)
    error = abs(w2 - w1)

    return w1, w2, error


def main():
    # Ignora a primeira linha
    _ = stdin.readline()

    values = [int(v) for v in stdin.readline().split()]
    m1, m2 = lb, ub = min(values), max(values)
    w1, w2, error = values_and_error(lb, ub, values)

    # Busca Ternária
    while error > EPS:
        third = (ub - lb) / 3
        m1 = lb + third
        m2 = ub - third

        w1, w2, error = values_and_error(m1, m2, values)

        if w2 > w1:
            ub = m2
        else:
            lb = m1

    x = (m1 + m2) / 2
    print(f"{weakness(x, values):.6f}")


if __name__ == "__main__":
    main()

File: ternary-search/codeforces-578-c/solution.cpp
#include <algorithm>
#include <climits>
#include <cstdio>
#include <cstdlib>

#define EPS 1e-6

double weekness(double x, double values[], int limit) {
  double max = std::abs(values[0] - x);
  double running_sum_pos = values[0] - x;
  double running_sum_neg = values[0] - x;

  for (int i = 1; i < limit; i++) {
    running_sum_pos = std::max(values[i] - x, values[i] - x + running_sum_pos);
    running_sum_neg = std::min(values[i] - x, values[i] - x + running_sum_neg);
    max = std::max(std::max(running_sum_pos, -running_sum_neg), max);
  }

  return max;
}

int main() {
  int n;
  double a[200000], lb = INT_MAX, ub = INT_MIN;

  // Lê o Input
  scanf("%d", &n);
  for (int i = 0; i < n; i++) {
    scanf("%lf", &a[i]);
    ub = std::max(ub, a[i]);
    lb = std::min(lb, a[i]);
  }

  // Busca ternária
  double m1, m2, w1, w2, error;
  do {
    m1 = lb + (ub - lb) / 3;
    m2 = ub - (ub - lb) / 3;

    w1 = weekness(m1, a, n);
    w2 = weekness(m2, a, n);

    if (w2 > w1)
      ub = m2;
    else
      lb = m1;

    error = std::abs(w2 - w1);
  } while (error > EPS);

  double x = (m1 + m2) / 2;
  printf("%.6lf\n", weekness(x, a, n));

  return 0;
}
